waku/v2/node/peer_manager/peer_manager.nim

```diff
@@ -3,75 +3,122 @@
 
-export peerstore, standard_setup
+export waku_peer_store
 
 declareCounter waku_peers_dials, "Number of peer dials", ["outcome"]
+declarePublicGauge waku_peers_errors, "Number of peer manager errors", ["type"]
 
 logScope:
   topics = "wakupeers"
 
 type
-  Connectedness* = enum
-    # NotConnected: default state for a new peer. No connection and no further information on connectedness.
-    NotConnected,
-    # CannotConnect: attempted to connect to peer, but failed.
-    CannotConnect,
-    # CanConnect: was recently connected to peer and disconnected gracefully.
-    CanConnect,
-    # Connected: actively connected to peer.
-    Connected
-  
-  ConnectionBook* = object of PeerBook[Connectedness]
-
-  WakuPeerStore* = ref object of PeerStore
-    connectionBook*: ConnectionBook
-
   PeerManager* = ref object of RootObj
     switch*: Switch
     peerStore*: WakuPeerStore
+    storage: PeerStorage
 
 const
   defaultDialTimeout = 1.minutes # @TODO should this be made configurable?
 
+####################
+# Helper functions #
+####################
+
+proc toPeerInfo(storedInfo: StoredInfo): PeerInfo =
+  PeerInfo.init(peerId = storedInfo.peerId,
+                addrs = toSeq(storedInfo.addrs),
+                protocols = toSeq(storedInfo.protos))
+
+proc insertOrReplace(ps: PeerStorage, peerId: PeerID, storedInfo: StoredInfo, connectedness: Connectedness) =
+  # Insert peer entry into persistent storage, or replace existing entry with updated info
+  let res = ps.put(peerId, storedInfo, connectedness)
+  if res.isErr:
+    warn "failed to store peers", err = res.error
+    waku_peers_errors.inc(labelValues = ["storage_failure"])
+
+proc dialPeer(pm: PeerManager, peerId: PeerID, 
+              addrs: seq[MultiAddress], proto: string,
+              dialTimeout = defaultDialTimeout): Future[Option[Connection]] {.async.} =
+  info "Dialing peer from manager", wireAddr = addrs[0], peerId = peerId
+
+  # Dial Peer
+  let dialFut = pm.switch.dial(peerId, addrs, proto)
+
+  try:
+    # Attempt to dial remote peer
+    if (await dialFut.withTimeout(dialTimeout)):
+      waku_peers_dials.inc(labelValues = ["successful"])
+      return some(dialFut.read())
+    else:
+      # @TODO any redial attempts?
+      debug "Dialing remote peer timed out"
+      waku_peers_dials.inc(labelValues = ["timeout"])
+
+      pm.peerStore.connectionBook.set(peerId, CannotConnect)
+      if not pm.storage.isNil:
+        pm.storage.insertOrReplace(peerId, pm.peerStore.get(peerId), CannotConnect)
+      
+      return none(Connection)
+  except CatchableError as e:
+    # @TODO any redial attempts?
+    debug "Dialing remote peer failed", msg = e.msg
+    waku_peers_dials.inc(labelValues = ["failed"])
+    
+    pm.peerStore.connectionBook.set(peerId, CannotConnect)
+    if not pm.storage.isNil:
+      pm.storage.insertOrReplace(peerId, pm.peerStore.get(peerId), CannotConnect)
+    
+    return none(Connection)
+
+proc primeFromStorage(pm: PeerManager) =
```

Comment:
What does this name mean? I find the "prime" verb a bit confusing

Reply:
I meant it in the sense of "prepare for usage". I think `loadFromStorage` may be less ambiguous.